materials_io/base.py

```diff
@@ -59,41 +58,57 @@ def parse(self, group: List[str], context: dict = None) -> dict:
             (dict): The parsed results, in JSON-serializable format.
         """
 
-    def parse_as_unit(self, files: List[str]) -> dict:
+    def parse_as_unit(self, files, context: dict = None) -> dict:
         """Parse a group of files and merge their metadata
 
         Used if each file in a group are parsed separately, but the resultant metadata
         should be combined after parsing.
 
         Args:
-            files ([str]): List of files to parse
+            files (str or [str]): Path(s) to parse
+            context (dict): Context about the files
         Returns:
-            (dict): Metadata summary from
+            (dict): Metadata summary from the files
         """
-
         # Initialize output dictionary
         metadata = {}
 
         # Loop over all files
-        for group in self.group(files):
+        for group in self.group(files, context):
             try:
-                record = self.parse(group)
+                record = self.parse(group, context)
             except Exception:
                 continue
-            metadata = dict_merge(metadata, record)
+            else:
+                metadata = dict_merge(metadata, record, append_lists=True)
         return metadata
 
-    def group(self, files: Iterable[str], context: dict = None) -> Iterator[Tuple[str, ...]]:
+    def group(self, files, context: dict = None) -> Iterator[Tuple[str, ...]]:
         """Identify a groups of files that should be parsed together
 
         Args:
-            files ([str]): List of paths, which could include both files and directories
+            files (str or [str]): Path(s), which could include both files and directories
             context (dict): Context about the files
         Yields:
             ((str)): Groups of files
         """
-
-        return zip(f for f in files if not os.path.isdir(f))
+        if isinstance(files, str):
+            files = [files]
+        # Clean paths
+        files = [os.path.abspath(os.path.expanduser(f)) for f in files]
+
+        # Default: Every file is in its own group, skipping symlinks
+        groups = []
+        for fn in files:
+            if os.path.isfile(fn):
+                groups.append((fn,))
+            # Recurse through directories (could also be done with os.walk(), especially
+            # for groups that may span directories)
+            elif os.path.isdir(fn):
```

Comment:
Something that occurred to me after looking at your code: Including the directory walking logic in `group` means that each implementation of `group` must also re-implement the directory-diving. 

Breaking out the core grouping logic could be to split off a function that "given my current directory, tell me the groups." That's pretty much what we had before, but with `group` performing the directory walking and having it calling a sub function.

Does that all line up? Should we revert back to `parse_directory` containing the walking code and `group` handling the contents of a directory (or whatever)?

Reply:
Hmm, or maybe the grouping operation could make groups from the groups generated by calling the superclass. But, that seems overly complicated.

Reply:
Yes, every implementation of `group` will have to walk the directories, in this architecture. However, breaking out the logic of "given only a list of files/a directory, group them" isn't going to have identical results. For example, we have concluded that some groups may span child directories. If the actual grouping logic is only given files, the subdirectory cannot be grouped. If the grouper is given the subdirectory, the files in the subdirectory will be double-grouped when the walker moves into the subdir.

For many cases, the default grouper will be sufficient. But when different groups should be made, the grouper's knowledge shouldn't be limited to only the current directory. It also isn't a significant ask of contributors to keep the directory walking in `group`, nor any added complexity compared to walking in a different method.

Reply:
I agree, keeping the grouping logic have the walking is OK. I agree it both isn't much to ask and not something many people will deal with on their own. To add your points: it provides also provides us the ability to specify to the grouper which directories it is allowed to dive into - rather than making always a "walk everything"